// octopus/lib/octopus_common.js
//This file implements common code between client and server.

var Octopus = { isServer:false };
try {
	Octopus = exports;
	Octopus.isServer = true;
} catch(e){}

(function() {
	var DEBUG_LOG_LEVEL = 0;

	var isPrimitive = function(obj) {
		if(obj == null || obj == undefined)
			return true;
		if(obj.constructor ===  Boolean)
			return true;
		if(obj.constructor ===  String)
			return true;
		if(obj.constructor ===  Number)
			return true;
		return false;
	};
	var isFunction = function(obj) {
		if(obj.constructor ===  Function)
			return true;
		return false;
	};
	var log = function(title, cont) {
		if (!DEBUG_LOG_LEVEL)
			return;
		try {
			if (Octopus.isServer)
				console.log( "   \x1b[36m" + title + " - \x1b[39m", cont);
			else
				console.log( "   " + title + " - ", cont);
		} catch(e) {
			cont_ = '{';
			for (key in cont) {
				if (!cont.constructor.prototype[key])
					cont_ += ' ' + key + ': ' + cont[key] + ', '
			}
			cont_ += '}';
			if (Octopus.isServer)
				console.log( "   \x1b[36m" + title + " - \x1b[39m" + cont_);
			else
				console.log( "   " + title + " - " + cont_);
		}
	};
	var tailCallCount = 0;
	var tailCall = function(func, arg) {
		tailCallCount++;
		if (tailCallCount > 1000) {
			tailCallCount = 0;
			setTimeout(function(){ func.apply(null, arg); }, 0);
		} else {
			func.apply(null, arg);
		}
	};

	var inner = {};
	Octopus.__inner__ = inner;

	inner.set_cps = function(r, prop, value, callback) {
		if (r instanceof RemoteObject)
			r.set_cps(prop, value, callback);
		else
			callback(r[prop] = value);
	};
	inner.get_cps = function(r, prop, callback) {
		if (r instanceof RemoteObject) {
			r.get_cps(prop, callback);
		} else {
			if (r[prop]) {
				callback(r[prop]);
			} else {
				if (r.get_cps) {
					r.get_cps(prop, callback);
				} else {
					callback(undefined);
				}
			}
		}
	};
	inner.each_cps = function(r, callback, k) {
		if (r instanceof RemoteObject)
			r.each_cps(callback, k);
		else
		{
			var keys = [];
			for (key in r) {
				if (key !== '__octopus_object_id__')
					keys.push(key);
			}
			var continue_k = function() {
				if (keys.length === 0) {
					k();
				} else {
					key = keys.pop();
					callback(key, k, continue_k);
				}
			};
			continue_k();
		}
	};

	var newApply = function(target, args) {
		var result, length = args.length;
		if (length == 0) result = new target();
		if (length == 1) result = new target(args[0]);
		if (length == 2) result = new target(args[0],args[1]);
		if (length == 3) result = new target(args[0],args[1],args[2]);
		if (length == 4) result = new target(args[0],args[1],args[2],args[3]);
		if (length == 5) result = new target(args[0],args[1],args[2],args[3],args[4]);
		if (length == 6) result = new target(args[0],args[1],args[2],args[3],args[4],args[5]);
		if (length == 7) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6]);
		if (length == 8) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7]);
		if (length == 9) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8]);
		if (length == 10) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9]);
		if (length == 11) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10]);
		if (length == 12) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11]);
		if (length == 13) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11],args[12]);
		if (length == 14) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11],args[12],args[13]);
		if (length == 15) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11],args[12],args[13],args[14]);
		if (length == 16) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11],args[12],args[13],args[14],args[15]);
		if (length == 17) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11],args[12],args[13],args[14],args[15],args[16]);
		if (length == 18) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11],args[12],args[13],args[14],args[15],args[16],args[17]);
		if (length == 19) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11],args[12],args[13],args[14],args[15],args[16],args[17],args[18]);
		if (length == 20) result = new target(args[0],args[1],args[2],args[3],args[4],args[5],args[6],args[7],args[8],args[9],args[10],args[11],args[12],args[13],args[14],args[15],args[16],args[17],args[18],args[19]);

		return result;
	};

	//TODO
	inner.new_cps = function(r, args, callback) {
		if (r === undefined || r === null)
			throw "No such function";
		// if r is remote, apply as remoteFunction.
		if (r.constructor === RemoteObject)
			r.new_cps(args, callback);
		else
		{
			// if r is native and userDefined function, just applying.
			if ( r.isUserDefined ) {
				args.push(callback);
				newApply(r, args);
			} else {
				// else, r is not userDefined so it is not translated. just translating args.
				var args_ = [];
				var k = function() {
					if (args.length == 0) {
						var result = newApply(r, args_);
						// TODO write back args, but because of few built-in function uses arguments writing, it is yet...
						callback(result);
					} else {
						var arg = args.shift();
						if (arg && arg.constructor === RemoteObject) {
							if (arg.isFunction) {
								// if arg is remoteFunction, wrapping anonymous function which calls origin as remote.
								var argFunc = function() {
									arg.apply_cps(null, arguments, function(){});
								};
								args_.push(argFunc);
								k();
							} else {
								// if arg is remote but not function, cloning arg from remote.
								arg.clone_cps(function(clone) {
									args_.push(clone);
									k();
								});
							}
						} else {
							// else, arg is native, just push to new args!
							args_.push(arg);
							k();
						}
					}
				}
				k();
			}
		}
	};

	inner.apply_cps = function(r, thisObj, args, callback) {
		if (r === undefined || r === null)
			throw "No such function";
		// if r is remote, apply as remoteFunction.
		if (r.constructor === RemoteObject)
			r.apply_cps(thisObj, args, callback);
		else
		{
			// if r is native and userDefined function, just applying.
			if ( r.isUserDefined ) {
				args.push(callback);
				r.apply(thisObj, args);
			} else {
				// else, r is not userDefined so it is not translated. just translating args.
				var args_ = [];
				var k = function() {
					if (args.length == 0) {
						var result = r.apply(thisObj, args_);
						// TODO write back args, but because of few built-in function uses arguments writing, it is yet...
						callback(result);
					} else {
						var arg = args.shift();
						if (arg && arg.constructor === RemoteObject) {
							if (arg.isFunction) {
								// if arg is remoteFunction, wrapping anonymous function which calls origin as remote.

								var argFunc = function() {
									arg.apply_cps(null, arguments, function(){});
								};
								args_.push(argFunc);
								k();
							} else {
								// if arg is remote but not function, cloning arg from remote.
								arg.clone_cps(function(clone) {
									args_.push(clone);
									k();
								});
							}
						} else {
							// else, arg is native, just push to new args!
							args_.push(arg);
							k();
						}
					}
				}
				k();
			}
		}
	};


	/**
	 * if
	 */
	inner.if_cps = function(condition, thenPart, elsePart, k) {
		if ( condition() )
			thenPart(k);
		else
			elsePart(k);
	};
	/**
	 * while(true)
	 */
	inner.loop_cps = function(loopBody, k) {
		var continue_k = function() {
			loopBody(k, continue_k);
		};
		continue_k();
	};

	exportedObjects = {};
	exportId = 1;
	/**
	 * RemoteObjectManager exists per client or server.
	 * @param clientSocket
	 * @param octopusServer
	 * @returns
	 */
	var RemoteObjectManager = function(socket, exportRoot) {
		this.socket = socket;
		this.objectTable = {};
		exportedObjects[0] = exportRoot;
		exportRoot.__octopus_object_id__ = 0;
		var that = this;

		// operation of receive message.
		socket.on('octopus_send_message', function (data) {
			that.receiveMessage(data);
		});
	};
	RemoteObjectManager.prototype.receiveMessage = function(data) {
		var socket = this.socket;
		var that = this;
		var returnResult = function(result) {
			log("return to other", result);
			socket.emit('octopus_return_'+data.objectId+'_'+data.invocationId, result);
		};
		var reciever = exportedObjects[data.objectId];
		log("Recieve message :" + data.type, data);
		log("  reciever:", reciever);
		switch (data.type) {
		case 'call':
			var args = [];
			for (var i = 0; i < data.args.length; i++)
				args.push( this.getRemoteObject(data.args[i]) );
			var thisObj = this.getRemoteObject(data.thisObj);
			inner.apply_cps(reciever, thisObj, args, function(result) { returnResult( that.exportObject(result) ); } );
			break;
		case 'new':
			var args = [];
			for (var i = 0; i < data.args.length; i++)
				args.push( this.getRemoteObject(data.args[i]) );
			inner.new_cps(reciever, args, function(result) { returnResult( that.exportObject(result) ); } );
			break;
		case 'set':
			var value = this.getRemoteObject(data.value);
			inner.set_cps(reciever, data.prop, value, function(result) { returnResult( that.exportObject(result) ); } );
			break;
		case 'get':
			inner.get_cps(reciever, data.prop, function(result) { returnResult( that.exportObject(result) ); } );
			break;
		case 'each':
			var keys = [];
			for (key in reciever) {
				if (key !== '__octopus_object_id__')
					keys.push(key);
			}
			returnResult(keys);
			break;
		case 'clone':
			returnResult(reciever);
			break;
		}
	};

	RemoteObjectManager.prototype.getRemoteObject = function(objData) {
		log("getRemoteObject", objData);
		// if objData is primitive, this is just a value!
		if (isPrimitive(objData))
			return objData;
		// if objData is native, return from exported table.
		if (objData.isNative === true)
			return exportedObjects[objData.remoteObjectId];
		// else, objData is remote, return as RemoteObject
		var objectId = objData.objectId;
		if (this.objectTable[objectId] !== undefined)
			return this.objectTable[objectId];
		return this.objectTable[objectId] = new RemoteObject(this, objectId, objData.isFunction);
	};
	// NativeObject -> remoteObjectId (with registering export table)
	RemoteObjectManager.prototype.getObjectId = function(object) {
		var objectId;
		if (object.__octopus_object_id__ === undefined) {
			objectId = exportId++;
			exportedObjects[objectId] = object;
			object.__octopus_object_id__ = objectId;
		}
		else {
			objectId = object.__octopus_object_id__;
		}
		log("getObjectId", objectId);
		return objectId;
	};
	// NativeObject -> ReturnMessage
	RemoteObjectManager.prototype.exportObject = function(object) {
		// if object is primitive, just return value.
		if (isPrimitive(object))
			return object;
		// if object is RemoteObject in sent PC, it is native and return a RemoteObjectId
		if (object.constructor === RemoteObject) {
			if (object.manager === this) {
				return { isNative:true, remoteObjectId:object.objectId };
			}
		}
		// else, get object id and return.
		var objectId = this.getObjectId(object);
		return { isNative:false, objectId:objectId, isFunction:(object.constructor === Function) };
	};

	inner.RemoteObjectManager = RemoteObjectManager;

	/**
	 * A constructor of RemoteObject.
	 * @param clientManager
	 * @param objectId
	 * @returns
	 */
	var RemoteObject = function(manager, objectId, isFunction) {
		this.manager = manager;
		this.objectId = objectId;
		this.invocationId = 0;
		this.isFunction = isFunction;
	};
	RemoteObject.prototype.sendMessage = function(msgObj, k) {
		var manager = this.manager;
		var socket = manager.socket;
		var invocationId = this.invocationId++;
		msgObj.invocationId = invocationId;
		msgObj.objectId = this.objectId;
		log("sendMessage", msgObj);
		socket.emit('octopus_send_message', msgObj);
		socket.on('octopus_return_'+this.objectId+'_'+invocationId, k);
	};
	RemoteObject.prototype.get_cps = function(prop, k) {
		var manager = this.manager;
		this.sendMessage({type:'get', prop:prop}, function(result) {
			if (k !== undefined) {
				k( manager.getRemoteObject(result) );
			}
		});
	};
	RemoteObject.prototype.set_cps = function(prop, value, k) {
		var manager = this.manager;
		this.sendMessage({type:'set', value:this.manager.exportObject(value), prop:prop}, function(result) {
			if (k !== undefined) {
				k( manager.getRemoteObject(result) );
			}
		});
	};

	RemoteObject.prototype.clone_cps = function(k) {
		this.sendMessage({type:'clone'}, function(result) {
			if (k !== undefined) {
				k( result );
			}
		});
	};
	RemoteObject.prototype.each_cps = function(callback, k) {
		this.sendMessage({type:'each'}, function(keys) {
			var continue_k = function() {
				if (keys.length == 0) {
					k();
				} else {
					key = keys.pop();
					callback(key, k, continue_k);
				}
			};
			continue_k();
		});
	};
	RemoteObject.prototype.apply_cps = function(thisObj, args, k) {
		var manager = this.manager;
		var args_ = [];
		for (var i = 0; i < args.length; i++) {
			args_.push( this.manager.exportObject(args[i]) );
		}
		this.sendMessage({type:'call', args:args_, thisObj:this.manager.exportObject(thisObj) }, function(result) {
			if (k !== undefined) {
				k( manager.getRemoteObject(result) );
			}
		});
	};

	RemoteObject.prototype.new_cps = function(args, k) {
		var manager = this.manager;
		var args_ = [];
		for (var i = 0; i < args.length; i++) {
			args_.push( this.manager.exportObject(args[i]) );
		}
		this.sendMessage({type:'new', args:args_}, function(result) {
			if (k !== undefined) {
				k( manager.getRemoteObject(result) );
			}
		});
	};

})();
